plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,122 @@
+#include "interface.h"
+#include "local_file.hpp"
+
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <fstream>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include <spdlog/spdlog.h>
+
+namespace
+{
+	std::string file_path;
+	std::vector<UserProfile> profile_list;
+	std::shared_mutex list_mutex;
+} // anonymous namespace
+
+auto init(const nlohmann::json& _config) -> void {
+	const auto path{_config.find("file_path")};
+	if (path == std::end(_config)) {
+		// handle bad case
+		spdlog::error("{}: Unable to find 'file_path' in provided config.", __func__);
+	}
+
+	// Save file string
+	file_path = *path;
+
+	// Parse into json object
+	spdlog::trace("{}: openning file_path [{}].", __func__, file_path);
+	update();
+}
+
+auto update() -> void {
+	try {
+		std::unique_lock update_lock{list_mutex, std::defer_lock};
+
+		std::ifstream file{file_path};
+		spdlog::debug("{}: Opened file [{}]", __func__, file_path);
+
+		auto json_data{nlohmann::json::parse(file)};
+
+		// Create temp list to be swapped in later
+		std::vector<UserProfile> temp_list;
+		temp_list.reserve(json_data.size());
+
+		// Process into list:
+		auto base_itter{json_data.items()};
+		std::transform(std::begin(base_itter), std::end(base_itter), std::back_inserter(temp_list), [](const auto& _itter) -> UserProfile {
+			return {.irods_user_name = _itter.key(), .attributes = _itter.value()};
+		});
+
+		// Move new item list to old one
+		if (update_lock.try_lock()) {
+			profile_list = std::move(temp_list);
+		}
```

Comment:
Yes.

Unless the number one goal is to make sure that the internal state is updated ASAP, I think `try_lock()` is fine.
I haven't 100% though through what it would mean if we just `lock()` though.

Any preferences?

Reply:
What you have is good.

I really dislike that the updated file contents is thrown away though. Adding some logic to only do the update when a change is detected (i.e. the mtime is different) would reduce the number of disk accesses.

Thoughts?

Reply:
Done. Only one thread can update at a time as well.